### ISSUE TITLE
need definition of which pseudo-classes can follow :part() that is sufficient to produce interoperability

### ISSUE THREAD

**Issue by dbaron** [w3c/csswg-drafts#10787]
[The spec](https://drafts.csswg.org/css-shadow-parts-1/#part) for `::part()` says:

> The [::part()](https://drafts.csswg.org/css-shadow-parts-1/#selectordef-part) pseudo-element can take additional pseudo-classes after it, such as x-button::part(label):hover, but never matches the [structural pseudo-classes](https://drafts.csswg.org/selectors-4/#structural-pseudo-classes) or any other pseudo-classes that match based on tree information rather than local element information.

This wording probably isn't precise enough to lead all implementors to classify every pseudo-class (works after `::part()` or not) in the same way.  I just added [a test](https://wpt.fyi/results/css/css-shadow-parts/pseudo-classes-after-part.html) to WPT (see also [the results from the PR](https://wpt.fyi/results/css/css-shadow-parts/pseudo-classes-after-part.html?label=pr_head&max-count=1&pr=47795) if the main results haven't populated yet) that exercises a bunch of these cases.  (See also #10786 for a separate issue that relates to the same test.  And also see #9795 for a case where people previously disagreed about this classification.)

We probably need a clearer definition of which selectors do or don't work here, and it probably needs to somehow end up in the way that we define pseudo-classes.

**Comment by dbaron** [comment]
Perhaps a special case:  I think it should probably be clear that `:has()` isn't allowed; it's probably not testable today but it could be in the future if we introduce something equivalent to what was allowed in early `:has()` drafts by `:has(:scope:hover)` (which is roughly equivalent to `:is(:hover)`).

**Comment by tabatkins** [comment]
Agreed.

I think the definitely-disallowed ones are:
* the structural pseudo-classes, as already disallowed
* `:has()`

Slightly uncertain ones that I think should still be allowed:
* :lang() and :dir() (we already resolved to allow them)
* the `:*-within` pseudos (`:target-within`, `:focus-within`
* `:picture-in-picture`

And I think the rest are all clearly fine.

I made this list by looking over <https://drafts.csswg.org/indexes/#selectors> which gets selectors from across the spec landscape, so should be pretty complete (anything it's missing are just mis-tagged).

**Comment by dbaron** [comment]
I think also probably disallowed (as structural) should be `:scope`, `:host-context()` and `:host`.

**Comment by tabatkins** [comment]
I'd considered them, but left them out as I figured they could never match. But you're right, for clarity they should be disallowed explicitly.

**Comment by dbaron** [comment]
> I made this list by looking over https://drafts.csswg.org/indexes/#selectors which gets selectors from across the spec landscape, so should be pretty complete (anything it's missing are just mis-tagged).

I checked the list in the index against Chrome's implementation.  The only pseudo-elements and pseudo-classes I found that aren't in the list, aren't prefixed, and don't work only on something else that's prefixed (like `::-webkit-scrollbar`) are:
* `:window-inactive` (shipping)
* [`:xr-overlay`](https://www.w3.org/TR/webxr-dom-overlays-1/#css-pseudo-class) (shipping)
* [`:granted`](https://github.com/WICG/PEPC/blob/main/explainer.md) (not enabled yet)
* [`::search-text`](https://github.com/w3c/csswg-drafts/issues/3812) (not enabled yet, not sure of status)
* [`::column`](https://github.com/flackr/carousel/blob/main/fragmentation/README.md),  [`::scroll-next-button`, and `::scroll-previous-button`](https://github.com/flackr/carousel/tree/main/scroll-button) (early carousel experiments, not enabled yet)

I don't think there are any mistagged specs in the CSS WG for any of these, though maybe `:xr-overlay` should show up in the index.

**Comment by schenney-chromium** [comment]
> I checked the list in the index against Chrome's implementation. The only pseudo-elements and pseudo-classes I found that aren't in the list, aren't prefixed, and don't work only on something else that's prefixed (like `::-webkit-scrollbar`) are:
> 
... 
> * [`::search-text`](https://github.com/w3c/csswg-drafts/issues/3812) (not enabled yet, not sure of status)

This one is in the prototype phase in Chrome but I would very much expect it to behave the same as ::selection and other highlights. Maybe runtime enabled status affects the behavior?

**Comment by tabatkins** [comment]
> :window-inactive (shipping)

Oh, MDN just documents this as a Moz extension only. If it's unprefixed and shipping it needs to show up in a spec.

Anyway, yeah, it should obviously match, it doesn't expose anything.

> [:xr-overlay](https://www.w3.org/TR/webxr-dom-overlays-1/#css-pseudo-class) (shipping)

This was indeed mistagged; Bikeshed doesn't infer CSS definitions from the text by default (the option is turned on in the CSSWG). I've submitted a PR to fix it.

This seems fine to match too, it's element state.

> [:granted](https://github.com/WICG/PEPC/blob/main/explainer.md?rgh-link-date=2024-09-22T20%3A39%3A28Z) (not enabled yet)

Same, this is element state, it should match.

**Comment by tabatkins** [comment]
For the pseudo-elements, I agree with Schenney that ::search-text should be treated identical to ::selection. Do we already have a term covering that category of things? If not, we should add one as it's probably useful to be able to refer to them as a group.

For the carousel pseudos, I think they should be supported. They're morally equivalent to ::before/::after for this purpose.

**Comment by schenney-chromium** [comment]
::search-text and ::selection are part of the Highlight Pseudos family. Also includes ::highlight(...), ::target-text, ::spelling-error and ::grammar-error.

**Comment by LeaVerou** [comment]
Shouldn’t `:state()` be one of these pseudo-classes?

**Comment by tabatkins** [comment]
Yeah, it's also mismarked and thus didn't show up in the list. Just discussed this with WHATWG folks, they'd prefer Selectors actually be the "definition" of :state(), and just defer to HTML for behavior, like we do with many other host-lang-specific things.

But yes, it should absolutely be allowed.

**Comment by tabatkins** [comment]
Ah, so `:state()` was actually marked up just fine, it just lives in Selectors 5.  It definitely shows up in the index, tho, so it fell under the "everything else is clearly allowed" category I gave.

**Comment by SebastianZ** [comment]
The [time-dimensional pseudo-classes](https://drafts.csswg.org/selectors-4/#time-pseudos) `:current` (and its functional notation `:current()`), `:past`, and `:future` sound like structural pseudo-classes to me. So I guess they should be excluded, right?

Sebastian
